File: backend/NodeJS/repositories/userRepository.js
const conn = require('../connection/connection');

const userRepository = {};

userRepository.getUserByEmail = (email) => {
    const query = `SELECT * FROM users WHERE email = ?`;
    return new Promise((resolve, reject) => {
        conn.query(query, [email], (err, result) => {
            if (err) return reject(err);
            resolve(result[0]);
        });
    });
};

userRepository.createUser = (firstname, lastname, email, password) => {
    const query = `INSERT INTO users(firstname, lastname, email, password) VALUES(?, ?, ?, ?)`;
    return new Promise((resolve, reject) => {
        conn.query(query, [firstname, lastname, email, password], (err) => {
            if (err) return reject(err);
            resolve();
        });
    });
};

userRepository.getUserByEmailAndPassword = (email, password) => {
    const query = `SELECT * FROM users WHERE email = ? AND password = ?`;
    return new Promise((resolve, reject) => {
        conn.query(query, [email, password], (err, result) => {
            if (err) return reject(err);
            resolve(result[0]);
        });
    });
};

userRepository.getAllUsers = () => {
    const query = `SELECT id, firstname, lastname, email, password, randomcode FROM users`;
    return new Promise((resolve, reject) => {
        conn.query(query, (err, result) => {
            if (err) return reject(err);
            resolve(result);
        });
    });
};

userRepository.getUserById = (userId) => {
    const query = `SELECT * FROM users WHERE id = ?`;
    return new Promise((resolve, reject) => {
        conn.query(query, [userId], (err, result) => {
            if (err) return reject(err);
            resolve(result[0]);
        });
    });
};

userRepository.updatePassword = (userId, newPassword) => {
    const query = 'UPDATE users SET password = ? WHERE id = ?';
    return new Promise((resolve, reject) => {
        conn.query(query, [newPassword, userId], (err, result) => {
            if (err) {
                return reject(new Error(err.message));
            }
            resolve(result);
        });
    });
};

userRepository.setRandomCode = (email, randomCode) => {
    const query = `UPDATE users SET randomCode = ? WHERE email = ?`;
    return new Promise((resolve, reject) => {
        conn.query(query, [randomCode, email], (err) => {
            if (err) return reject(err);
            resolve();
        });
    });
};

userRepository.verifyRandomCode = (randomCode) => {
    const query = `SELECT * FROM users WHERE randomCode = ?`;
    return new Promise((resolve, reject) => {
        conn.query(query, [randomCode], (err, result) => {
            if (err) return reject(err);
            resolve(result.length > 0);
        });
    });
};

userRepository.clearRandomCode = (randomCode) => {
    const query = `UPDATE users SET randomCode = NULL WHERE randomCode = ?`;
    return new Promise((resolve, reject) => {
        conn.query(query, [randomCode], (err) => {
            if (err) return reject(err);
            resolve();
        });
    });
};

userRepository.resetPassword = (email, newPassword) => {
    const query = `UPDATE users SET password = ? WHERE email = ?`;
    return new Promise((resolve, reject) => {
        conn.query(query, [newPassword, email], (err) => {
            if (err) return reject(err);
            resolve();
        });
    });
};

module.exports = userRepository;

File: backend/app/repositories/user_repository.py
from app.infrastructure.db_connection_manager import DbConnectionManager
from mysql.connector import Error
from typing import Optional
import pymysql.cursors


class UserRepository:
    def __init__(self):
        self.connection = None

    def connect(self):
        """Establishes the database connection using DbConnectionManager."""
        self.connection = DbConnectionManager.get_connection()

    def get_user_by_email(self, email: str) -> Optional[dict]:
        """Fetches a user by their email."""
        self.connect()

        try:
            if not self.connection:
                print("Not connected to the database.")
                return None

            with self.connection.cursor(pymysql.cursors.DictCursor) as cursor:
                cursor.execute("SELECT * FROM users WHERE email = %s", (email,))
                user = cursor.fetchone()

            return user

        except Error as e:
            print(f"Error: {e}")
            return None

    def create_user(self, firstname: str, lastname: str, email: str, password: str) -> None:
        """Inserts a new user into the users table."""
        self.connect()

        try:
            if not self.connection:
                print("Not connected to the database.")
                return

            with self.connection.cursor() as cursor:
                cursor.execute(
                    """
                    INSERT INTO users (firstname, lastname, email, password)
                    VALUES (%s, %s, %s, %s)
                    """,
                    (firstname, lastname, email, password)
                )
                self.connection.commit()
                print("User created successfully.")

        except Error as e:
            print(f"Error: {e}")

    def get_user_by_email_and_password(self, email: str, password: str) -> Optional[dict]:
        """Fetches a user by their email and password."""
        self.connect()

        try:
            if not self.connection:
                print("Not connected to the database.")
                return None

            with self.connection.cursor(pymysql.cursors.DictCursor) as cursor:
                cursor.execute("SELECT * FROM users WHERE email = %s AND password = %s", (email, password))
                user = cursor.fetchone()

            return user

        except Error as e:
            print(f"Error: {e}")
            return None
